src/modbus/ModbusBridge.cpp

```diff
@@ -134,535 +158,553 @@ ModbusBridge::~ModbusBridge()
+    auto values = StringUtils::tokenize(mapping.getValue(), ",");
     for (auto& value : values)
     {
         configurationValues.emplace_back(value);
     }
 
-    return ConfigurationItem(configurationValues, modbusRegisterMapping.getReference());
+    return ConfigurationItem(configurationValues, mapping.getReference());
 }
+// getConfigurationStatusFromCoil() - handles
+ConfigurationItem ModbusBridge::getConfigurationStatusFromCoil(ModbusRegisterMapping& mapping)
+{
+    bool value;
+    if (!m_modbusClient.readCoil(mapping.getSlaveAddress(), mapping.getAddress(), value))
+    {
+        LOG(ERROR) << "ModbusBridge: Unable to read coil on address '" << mapping.getAddress() << "'";
 
-wolkabout::DeviceStatus::Status ModbusBridge::getDeviceStatus(const std::string& /* deviceKey */)
+        mapping.setValid(false);
+        return ConfigurationItem(std::vector<std::string>(), "");
+    }
+    mapping.update(value);
+    return ConfigurationItem({mapping.getValue()}, mapping.getReference());
+}
+// getConfiguration() - decides on the handler to handle the status request
+std::vector<ConfigurationItem> ModbusBridge::getConfiguration(const std::string& deviceKey)
 {
-    return m_modbusClient.isConnected() ? DeviceStatus::Status::CONNECTED : DeviceStatus::Status::OFFLINE;
+    std::vector<ConfigurationItem> configurations;
+    for (const auto& referenceMappingPair : m_configurationMappingByDevice[deviceKey])
+    {
+        auto& mapping = *referenceMappingPair.second;
+
+        switch (mapping.getRegisterType())
+        {
+        case ModbusRegisterMapping::RegisterType::COIL:
+            configurations.emplace_back(getConfigurationStatusFromCoil(mapping));
+            break;
+        case ModbusRegisterMapping::RegisterType::HOLDING_REGISTER_ACTUATOR:
+            configurations.emplace_back(getConfigurationStatusFromHoldingRegister(mapping));
+            break;
+        default:
+            throw std::logic_error("Invalid type of register for a configuration at reference '" +
+                                   referenceMappingPair.first + "'!");
+        }
+    }
+
+    return configurations;
+}
+
+// getDeviceStatus() - handles status request about a specific device
+wolkabout::DeviceStatus::Status ModbusBridge::getDeviceStatus(const std::string& deviceKey)
+{
+    return m_devicesStatusBySlaveAddress[getSlaveAddress(deviceKey)];
 }
 
+// methods for the running logic of modbusBridge
 void ModbusBridge::start()
 {
+    LOG(DEBUG) << "Starting ModbusBridge.";
     if (m_readerShouldRun)
     {
         return;
     }
 
     m_readerShouldRun = true;
     m_reader = std::unique_ptr<std::thread>(new std::thread(&ModbusBridge::run, this));
+    LOG(DEBUG) << "Started ModbusBridge.";
 }
 
 void ModbusBridge::stop()
 {
     m_readerShouldRun = false;
-    m_reader->join();
+    if (m_reader->joinable())
+    {
+        m_reader->join();
+    }
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
```

Comment:
why sleep?

Reply:
That is the only way to ensure that offline messages get to the platform after the pending data messages, since they travel slower. WolkGatewayModuleSDK-Cpp doesn't provide any way to see if there's messages being sent/have arrived. We just need to wait for 2s, hope they all arrived there, and then notify of status. This has been brought up to the backend, where they will implement a solution for this.